KAFKA-10162; Use Token Bucket algorithm for controller mutation quota (KIP-599, Part III)

dajac opened the pull request:
Based on the discussion in https://github.com/apache/kafka/pull/9072, I have put together an alternative way. This one does the following:
* Instead of changing the implementation of the Rate to behave like a Token Bucket, it actually use two different metrics: the regular Rate and a new Token Bucket. The latter is used to enforce the quota.
* The Token Bucket algorithm uses the rate of the quota as the refill rate for the credits and compute the burst based on the number of samples and their length (# samples * sample length * quota).
* The Token Bucket algorithm used can go under zero in order to handle unlimited burst (e.g. create topic with a number of partitions higher than the burst). Throttling kicks in when the number of credits is under zero.
* The throttle time is computed as credits under zero / refill rate (or quota).
* Only the controller mutation uses it for now.
* The remaining number of credits in the bucket is exposed with the `tokens` metrics per user/clientId.

The code can be improved and refactored. I just wanted to get out quickly to get feedback about the approach.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

dajac committed:
KAFKA-10162; Make the rate based quota behave more like a Token Bucket (KIP-599, Part III)

junrao commented on clients/src/main/java/org/apache/kafka/common/metrics/stats/TokenBucket.java:
```diff
@@ -0,0 +1,83 @@
+        if (config.quota() == null)
+            return;
+        final double quota = config.quota().bound();
+        final double burst = (config.samples() - 1) * convert(config.timeWindowMs()) * quota;
```
Should burst be computed from `#samples` or `#samples - 1` ?

apovzner replied:
If we want the burst to be more similar to original behavior, it seems like this should be `#samples`. With the current implementation, we can do 1 unit of work in the oldest window and then accept a burst right at the end of the last (not full yet) sample. Which means that the max burst size is almost at #samples * quota (if sample = 1 sec, quota is in units/second). Does this sound right to you?

Also, I think we should take into account `config.timeWindowMs`, because it could be something other than 1 second.

dajac replied:
That's a very good question. I use `#samples - 1` here because we do so in `Rate#windowSize` to compute the total window so it is probably closer like this.

@apovzner `config.timeWindowMs` is already in the formula or did I misunderstand your comment?

apovzner replied:
hm.. maybe I just did not notice `config.timeWindowMs`. All good now.

junrao commented on core/src/main/scala/kafka/server/ControllerMutationQuotaManager.scala:
```diff
@@ -131,6 +133,16 @@ class PermissiveControllerMutationQuota(private val time: Time,
+
+  def throttleTime(e: QuotaViolationException, timeMs: Long): Long = {
+    e.metric().measurable() match {
+      case _: TokenBucket => Math.round(-e.value() * e.bound())
```
Hmm, assuming bound is the per sec rate, it seems that the throttleTimeMs should be `-e.value() / e.bound * 1000`?

dajac replied:
Yes, that's right.

junrao commented on clients/src/main/java/org/apache/kafka/common/metrics/stats/TokenBucket.java:
```diff
@@ -0,0 +1,83 @@
+
+    public TokenBucket(TimeUnit unit) {
+        this.unit = unit;
+        this.credits = 0;
```
Should we start with 0 credit or the full burst credits? The benefit of the latter is that during initialization, the requests won't be throttled as much due to a cold start.

apovzner replied:
I also think it's better to start with burst size credits. We would normally create the sensor on the first byte/request/etc. and before that, the <user,client> is idle which means they were "accumulating credits".

dajac replied:
The burst is not know when the `TokenBucket` is constructed as we must get the `MetricConfig` at least once to initialise it correctly. Note that, it will be initialised to the maximum burst when we refill the bucket when either `measure` or `record` is called for the first time. That works because `lastUpdateMs` is `0` so the delta between `now` and it will make the bucket full. So, the bucket actually stats from the maximum burst.

apovzner replied:
ah right, `lastUpdateMs` will make sure that bucket would be full on the first `record()`.

junrao reviewed (commented):
@dajac : Thanks for the PR. Took a quick look. It does look like this approach is less intrusive and probably easier to understand. A few quick comments below.

apovzner commented on clients/src/main/java/org/apache/kafka/common/metrics/Sensor.java:
```diff
@@ -223,8 +229,14 @@ public void checkQuotas(long timeMs) {
-                    if (!quota.acceptable(value)) {
-                        throw new QuotaViolationException(metric, value, quota.bound());
+                    if (metric.measurable() instanceof TokenBucket) {
+                        if (value <= 0) {
```
we probably just want `value < 0` check here, right? Otherwise the throttle time will be 0 anyways.

dajac replied:
Yes, that makes sense.

apovzner commented:
I also took a quick look and I agree that this implementation is cleaner and definitely easier to understand.

dajac committed:
refactor, add tests, address comments

dajac commented:
@junrao @apovzner I have updated the PR to address your comments. I have also refactored the code a bit and updated all the tests of the controller quota manager to work with the Token Bucket.

From a public api perspective, we would have a new metric per user/clientId named: `tokens`. It represents the number of tokens in the bucket.

Please, pay close attention to the unit tests of the controller mutation quota manager. Most of the throttle times have been updated there to match the token bucket usage. This is due to the fact that the token bucket does not have to wait until old samples are purged to get it amount of tokens updated. It is updated continuously.

I wonder if we should use by default strait away of put it behind a config flag to let user opt in. It does not matter much for our new quota but that may matter if/when we migrate existing quotas. What do you think?

junrao commented on clients/src/main/java/org/apache/kafka/common/metrics/stats/TokenBucket.java:
```diff
@@ -0,0 +1,87 @@
+import org.apache.kafka.common.metrics.MeasurableStat;
+import org.apache.kafka.common.metrics.MetricConfig;
+
+public class TokenBucket implements MeasurableStat {
```
Could we add a high level description of the class?

dajac replied:
Sure.

junrao commented on clients/src/main/java/org/apache/kafka/common/metrics/stats/TokenBucket.java:
```diff
@@ -0,0 +1,87 @@
+    }
+
+    private double burst(final MetricConfig config) {
+        return (config.samples() - 1) * convert(config.timeWindowMs()) * config.quota().bound();
```
Rate actually allows the windowSize to be close to the full samples * perSampleWindow. The logic around `config.samples() - 1` is just to make sure the windowSize contains at least that many full windows. So, to match that behavior, it seems that burst should use `config.samples()`.

dajac replied:
Ack. I misunderstood it. `config.samples()` sounds good to me.

junrao commented on clients/src/main/java/org/apache/kafka/common/metrics/stats/TokenBucket.java:
```diff
@@ -0,0 +1,87 @@
+    }
+
+    private double convert(final long timeMs) {
+        switch (unit) {
```
This code is duplicated from Rate. Could we reuse it somehow?

dajac replied:
Sure, makes sense.

junrao commented on core/src/main/scala/kafka/server/ControllerMutationQuotaManager.scala:
```diff
@@ -131,6 +133,22 @@ class PermissiveControllerMutationQuota(private val time: Time,
+   * Basically, if a value < 0 is observed, the time required to bring it to zero is
+   * -value / refill rate (quota bound) * 1000.
+   */
+  def throttleTime(e: QuotaViolationException, timeMs: Long): Long = {
```
throttleTime => throttleTimeMs ?

junrao commented on core/src/main/scala/kafka/server/ControllerMutationQuotaManager.scala:
```diff
@@ -156,6 +174,24 @@ class ControllerMutationQuotaManager(private val config: ClientQuotaManagerConfi
       quotaMetricTags.asJava)
   }
 
+  protected def clientTokenBucketMetricName(quotaMetricTags: Map[String, String]): MetricName = {
```
Could this be private?

dajac replied:
Yes.

junrao commented on clients/src/test/java/org/apache/kafka/common/metrics/SensorTest.java:
```diff
@@ -209,4 +212,95 @@ public void shouldReturnPresenceOfMetrics() {
+
+    private void strictRecord(Sensor sensor, double value, long timeMs) {
+        synchronized (sensor) {
+            sensor.checkQuotas(timeMs);
```
sensor.record() always calls checkQuotas(). Why do we need to call it explicitly here?

dajac replied:
In the above two tests, I simulate a "strict quotas" in the sense that recording is not allowed if the quota is already violated. Therefore, I check it before recording the value.

junrao commented on clients/src/test/java/org/apache/kafka/common/metrics/SensorTest.java:
```diff
@@ -209,4 +212,95 @@ public void shouldReturnPresenceOfMetrics() {
+
+        Mockito.when(stat1.measure(stat1Config, 2)).thenReturn(2.0);
+        Mockito.when(stat2.measure(stat2Config, 2)).thenReturn(2.0);
+        sensor.checkQuotas(2);
```
Here, we are just verifying there is no quota exception?

dajac replied:
No. This test actually reproduce a bug that I have found. Basically, a stat can be added to the Sensor with a MetricsConfig but the Sensor was not using the provided one when recording a value but was using the one of the Sensor all the time. This test verifies that the correct config is used both for recording and measuring via calling checkQuota.

junrao commented on clients/src/test/java/org/apache/kafka/common/metrics/TokenBucketTest.java:
```diff
@@ -0,0 +1,93 @@
+        // Expect 100 credits at T
+        assertEquals(100, tk.measure(config, time.milliseconds()), 0.1);
+
+        // Record 60 at T, expect 100 credits
```
We are recording -60.

junrao commented on clients/src/test/java/org/apache/kafka/common/metrics/TokenBucketTest.java:
```diff
@@ -0,0 +1,93 @@
+        tk.record(config, 60, time.milliseconds());
+        assertEquals(40, tk.measure(config, time.milliseconds()), 0.1);
+
+        // Advance by 2s, record 60, expect 100 credits
```
We are recording -60.

junrao reviewed (commented):
@dajac : Thanks for the updated PR. Looks good overall. Just a few comments below.

dajac committed:
Address Jun's comments

dajac commented:
@junrao Thanks for your comments. I just pushed an update which addressed them.

junrao commented on clients/src/main/java/org/apache/kafka/common/metrics/stats/TokenBucket.java:
```diff
@@ -0,0 +1,80 @@
+ * the maximum number of credits of the bucket is defined by
+ * {@link MetricConfig#samples() * MetricConfig#timeWindowMs() * Quota#bound()}.
+ *
+ * The quota is considered as exhausted when the amount of remaining credits in the bucket
```
Could we document how this quota behaves differently from existing quota?

junrao commented on clients/src/test/java/org/apache/kafka/common/metrics/TokenBucketTest.java:
```diff
@@ -0,0 +1,93 @@
+    @Test
+    public void testRecord() {
+        // Rate  = 5 unit / sec
+        // Burst = 2 * (11 - 1) = 20 units
```
we have 10 samples now.

junrao commented on clients/src/test/java/org/apache/kafka/common/metrics/TokenBucketTest.java:
```diff
@@ -0,0 +1,93 @@
+    @Test
+    public void testUnrecord() {
+        // Rate  = 5 unit / sec
+        // Burst = 2 * (11 - 1) = 20 units
```
we have 10 samples now.

junrao commented on clients/src/test/java/org/apache/kafka/common/metrics/SensorTest.java:
```diff
@@ -209,4 +212,95 @@ public void shouldReturnPresenceOfMetrics() {
+        final Sensor sensor = metrics.sensor("sensor", new MetricConfig()
+            .quota(Quota.upperBound(2))
+            .timeWindow(1, TimeUnit.SECONDS)
+            .samples(11));
```
Is the test based on 10 samples?

dajac replied:
The config is correct. 11 samples. With the few samples in the test, the total window is actually 10s. This is why I use 10 in the formulas.

junrao reviewed (commented):
@dajac : Thanks for the updated PR. Just a few more minor comments below.

junrao commented:
ok to test

dajac committed:
address Jun's comments

dajac commented:
@junrao Thanks. I have updated the PR.

junrao commented on clients/src/main/java/org/apache/kafka/common/metrics/stats/TokenBucket.java:
```diff
@@ -0,0 +1,101 @@
+ *
+ * Token Bucket vs Rate based Quota:
+ * The current sampled rate based quota does not cope well with bursty workloads. The issue is
+ * that a unique and large sample can hold the average above the quota and this until it is
```
"this until it is" doesn't quite parse.

dajac replied:
i have removed the "and this". Does it parse better?

junrao commented on clients/src/main/java/org/apache/kafka/common/metrics/stats/TokenBucket.java:
```diff
@@ -0,0 +1,101 @@
+ * - Quota (Q)   = 5
+ * - Samples (S) = 100
+ * - Window (W)  = 1s
+ * A burst of 560 brings the average rate (R) to 5.6 (560 / 100). The throttle time is computed as
```
"The throttle time " : I guess this is the expected throttle time?

dajac replied:
Yes, that is correct. I have added "expected" to be clearer.

junrao reviewed (commented):
@dajac : Thanks for the updated PR. A couple of more minor comments.

junrao commented:
ok to test

junrao commented:
retest this please

junrao commented:
test this please

dajac committed:
fixup

junrao reviewed (approved):
@dajac : Thanks for the latest update. LGTM

For your earlier question, if this works out well. We could just change existing quotas' implementation accordingly without necessarily a new flag to opt in.

Could you also update the KIP and the email thread with the newly added token metric to make sure there is no objection to this minor change?

junrao commented:
retest this please

junrao commented:
test this please

dajac commented:
Sure. I will update the KIP and the mailing list tomorrow. Thanks, Jun!

junrao commented:
The failed unit test seems unrelated.

@apovzner : Do you have any other comments on this? Thanks.

apovzner reviewed (approved):
Looks great!

dajac commented:
@junrao I have updated the KIP and posted an update to the thread in the mailing list.
